aptos-move/move-examples/data_structures/sources/smart_tree.move

```diff
@@ -0,0 +1,971 @@
+/// Type of large-scale search trees.
+///
+/// It internally uses BTree to organize the search tree data structure for keys. Comparing with
+/// other common search trees like AVL or Red-black tree, a BTree node has more children, and packs
+/// more metadata into one node, which is more disk friendly (and gas friendly).
+
+module aptos_std::smart_tree {
+    use aptos_std::bucket_table::{Self, BucketTable};
+    use std::option::{Self, Option};
+    use std::vector;
+
+    // Internal errors.
+    const E_INTERNAL: u64 = 0;
+    // The tree is not empty, and cannot be destroyed.
+    const E_TREE_NOT_EMPTY: u64 = 1;
+    // The tree is too big for insertion.
+    const E_TREE_TOO_BIG: u64 = 2;
+    // The provided parameter is invalid.
+    const E_INVALID_PARAMETER: u64 = 3;
+
+    const NULL_INDEX: u64 = 0;
+    const DEFAULT_ORDER : u8 = 32;
+    const MAX_NUM_NODES : u64 = 1048576;
+
+    /// A node of the SmartTree.
+    struct Node has drop, store {
+        // Whether this node is a leaf node.
+        is_leaf: bool,
+        // The node index of its parent node, or NULL_INDEX if it doesn't have parent.
+        parent: u64,
+        // The children of the nodes. (When the node is leaf node, all keys of the node is stored in children.max_key)
+        children: vector<Child>,
+        // The node index of its previous node at the same layer, or NULL_INDEX if it doesn't have a previous node.
+        prev: u64,
+        // The node index of its next node at the same layer, or NULL_INDEX if it doesn't have a previous node.
+        next: u64,
+    }
+
+    /// The metadata of a child of a node.
+    struct Child has copy, drop, store {
+        // The max key of it's child, or the key if the current node is a leaf node.
+        max_key: u64,
+        // The node index of it's child, or NULL_INDEX if the current node is a leaf node.
+        node_index: u64,
+    }
+
+    /// An iterator to iterate all keys in the SmartTree.
+    struct Iterator has copy, drop {
+        // The node index of the iterator pointing to.
+        node_index: u64,
+        // The child index of the iterator pointing to.
+        child_index: u64,
+        // The key of the iterator pointing to, not valid when the iterator is an end iterator.
+        key: u64,
+    }
+
+    /// The SmartTree data structure.
+    struct SmartTree<V> has store {
+        // The node index of the root node.
+        root_index: u64,
+        // Mapping of node_index -> node.
+        nodes: BucketTable<u64, Node>,
```

Comment:
It appears you are using BucketTable here to represent tree nodes, but how is it guaranteed that you aren't widely jumping around in the buckets to do e.g. a simple binary search? For example, you follow childs from a given node, but unfortunately each node on your search leads to loading an entire new bucket... There should be some affinity of buckets to traversal order, not sure if this is given?

Reply:
Good question!  I need everyone's opinion here. @davidiw @msmouse @junkil-park @lightmark @zekun000 @movekevin @vgao1996 

What my current opinion is:

for `nodes`, I'm now feeling we probably just want a Table/TableWithLength, given the size of a Node is only depending on the number of children it has, people can control it to make it align with our gas schedule.

for `entries`, maybe @lightmark 's `SmartTable`?

Reply:
Adding my 2cents here:

We've implemented something very similar to this SmartTree for ferum. Like @grao1991 mentioned, we just used a regular Table and relied on changing the size of the Node for "bucketing". For our use case, Nodes store sequential data so that made a lot of sense. 

For BucketTable to be useful, the buckets need to store Nodes that are related to one another.